### DISCUSSION TITLE
ComboBox changes value of bound data just before having its DataContext changed

### DISCUSSION THREAD

**Discussion by amSnoop** [AvaloniaUI/Avalonia#13488] [Q&A]
Breakdown:

CueEditorWindow is  UserControl whose data context contains an ObservableCollection of CueFileViewModels:

```csharp
[ObservableProperty]
private ObservableCollection<CueFileViewModel> cues = new();
```
Inside the control, there is an ItemsControl that binds to this list. Each item in the ItemsControl is inside of an expander that has an ComboBox in it. That ComboBox is bound to the CueFileViewModel.File.EndAction property:

```xaml
<ItemsControl Background="#090909" ItemsSource="{Binding Cues}">
				<ItemsControl.ItemTemplate>
					<DataTemplate>
						<Expander Focusable="False" IsExpanded="True" Classes.Error="{Binding IsError}" BorderBrush="White" Header="{Binding Header}" Foreground="#D345A4" HorizontalAlignment="Stretch">
							<Grid Grid.Row="2" Grid.Column="1" x:Name="DefaultCueDisplay">
								<Grid.RowDefinitions>
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
									<RowDefinition Height="*" />
								</Grid.RowDefinitions>
								<Grid.ColumnDefinitions>
									<ColumnDefinition Width="*" />
									<ColumnDefinition Width="*" />
								</Grid.ColumnDefinitions>

								<TextBlock Grid.Row="1" Text="File Path:" VerticalAlignment="Center" />
								<Grid ColumnDefinitions="20*, *,Auto" Grid.Row="1" Grid.Column="1">
									<TextBox Grid.Row="1" Grid.Column="0" HorizontalAlignment="Stretch" VerticalAlignment="Center" Text="{Binding File.FilePath}"/>
									<Button Focusable="False" Grid.Column="2" Content=". . ." Margin="2,0,0,0" FontSize="14" VerticalAlignment="Center" Padding="3,6" BorderThickness="1" BorderBrush="#444" Background="#333" Command="{Binding OpenFileDialog}" CommandParameter="{Binding $parent[Window]}"/>
								</Grid>

								<TextBlock Grid.Row="2" Text="Fade In:" VerticalAlignment="Center"/>
								<TextBox Grid.Row="2" Grid.Column="1" Width="50" HorizontalAlignment="Right" VerticalAlignment="Center" Text="{Binding File.FadeInDuration}" />

								<TextBlock Grid.Row="3" Text="Fade Out:" VerticalAlignment="Center"/>
								<TextBox Grid.Row="3" Grid.Column="1" Width="50" HorizontalAlignment="Right" VerticalAlignment="Center" Text="{Binding File.FadeOutDuration}" />
<!--RIGHT HERE-->
								<TextBlock Grid.Row="4" Text="Cue Ending Behavior:" VerticalAlignment="Center"/>
								<ComboBox Focusable="False" Grid.Row="4" Grid.Column="1" HorizontalAlignment="Stretch" VerticalAlignment="Center"  SelectedIndex="{Binding File.EndAction, Mode=TwoWay}" ItemsSource="{Binding $parent[local:CueEditorControl].DataContext.EndActionStrings}" />

								<TextBlock Grid.Row="5" Text="Duration:" VerticalAlignment="Center"  IsVisible="{Binding DurationVisibility}"/>
								<TextBox Grid.Row="5" Grid.Column="1" Width="50" HorizontalAlignment="Right" VerticalAlignment="Center" Text="{Binding File.Duration}" IsVisible="{Binding DurationVisibility}" />

								<TextBlock Grid.Row="5" Text="Volume:" VerticalAlignment="Center" IsVisible="{Binding VolumeVisibility}"/>
								<Grid Grid.Row="5" Grid.Column="1" ColumnDefinitions="20*, *,Auto" IsVisible="{Binding VolumeVisibility}">
									<Slider Focusable="False" HorizontalAlignment="Stretch" VerticalAlignment="Center" Maximum="100" Minimum="0" TickFrequency="1" Value="{Binding File.Volume}"/>
									<TextBox Grid.Column="2" Width="50" HorizontalAlignment="Right" VerticalAlignment="Center" Text="{Binding File.Volume, Mode=TwoWay}" />
								</Grid>
							</Grid>
						</Expander>
					</DataTemplate>
				</ItemsControl.ItemTemplate>
</ItemsControl>
```

When I set the value by selecting something from the ComboBox, everything goes smoothly, and it updates the relevent property in the model. The issue is that somewhere in this call:
            ```CueEditorWindow.DataContext = cueEditorViewModel;```
            
There is a call to the ComboBox to change its value, which gets propagated to the model, before the DataContext has been updated. The value being sent is -1 which shouldn't even be possible. 

I tried digging through the call stack to figure out where the actual change is being made, but I am not experienced enough to recognize the issue.

Another note is that this line from earlier in the file (not a part of that list of Cues or in the expander, but directly bound to `DataContext.Cue.CueEndAction` works perfectly:

```xaml
				<TextBlock Grid.Row="4" Text="Cue Ending Behavior:" VerticalAlignment="Center" Margin="0,10"/>
				<ComboBox Focusable="False" Grid.Row="4" Grid.Column="1" HorizontalAlignment="Stretch" VerticalAlignment="Center"  
						  SelectedIndex="{Binding Cue.CueEndAction, Mode=TwoWay}" ItemsSource="{Binding EndActionStrings}" />
```
This entire file can be found in [this repository](https://github.com/amSnoop/DeLight/blob/main/DeLight/Views/CueEditorControl.axaml)

As a side note, is there an easy way to remove all the excess indentation when copying from indented text?

Thanks.

**Comment by timunie** [comment]
Try to set selection after Items source. It can't select an item that's not yet present.

**Reply by amSnoop** [reply]
I'm not entirely sure if I understand what you mean, but if you meant for me to do this:
```xaml
<ComboBox Focusable="False" Grid.Row="4" Grid.Column="1" HorizontalAlignment="Stretch" VerticalAlignment="Center" 
ItemsSource="{Binding $parent[local:CueEditorControl].DataContext.EndActionStrings}" 
SelectedIndex="{Binding File.EndAction, Mode=TwoWay}" />
```
Then it had no effect. I should also specify that the issue is that it's setting the value on the *outgoing* data context, not the incoming one.

**Comment by timunie** [comment]
> As a side note, is there an easy way to remove all the excess indentation when copying from indented text?

Select all and [Shift] + [Tab]

**Comment by amSnoop** [comment]
I seem to have found a workaround, although I'm not entirely sure why it works or how it works.

You can see in this line that this is binding its ItemSource to it's parent's DataContext. I suspect that this search for items happens out of order with it setting its selected index causing the issue to appear. 

```xaml
<ComboBox Focusable="False" Grid.Row="4" Grid.Column="1" HorizontalAlignment="Stretch" VerticalAlignment="Center" 
ItemsSource="{Binding $parent[local:CueEditorControl].DataContext.EndActionStrings}" SelectedIndex="{Binding File.EndAction, Mode=TwoWay}" />
```
That's this list:
```csharp
        public static List<string> EndActionStrings { get; } = new() {
            "Loop" ,
            "Fade After End" ,
            "Fade Before End" ,
            "Freeze"
        };
```
If I change it to this:
```xaml
<ComboBox Focusable="False" Grid.Row="4" Grid.Column="1" HorizontalAlignment="Stretch" VerticalAlignment="Center" 
ItemsSource="{Binding EndActionStrings}" SelectedIndex="{Binding File.EndAction, Mode=TwoWay}" />
```
and place an identical copy of that list in the sub-viewmodels that that list binds its items to, then it works, with some notable exceptions:

- It did not work if it was static
- ~~It did not work if I placed it in the base CueFileViewModel class, only the Light and ScreenFileViewModel child classes.~~ Edit: I swear I tried this and it didn't work, but me reverting some files on accident led to this being tested again, and now it seems to work just fine.

Is this intended behavior? It seems weird ~~to me that it can't find it's parent's bound properties in time for setting the value, and it seems weird~~ that it can't access static variables from its own data context.

Thanks again.

**Reply by timunie** [reply]
Probably it doesn't find the parent, yeah. You can also give a name to your parent and bind to it via #root.DataContext....
